src/datatablerow/utils.jl

```diff
@@ -0,0 +1,190 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    dt::T                 # source data frame
+
+    ngroups::Int          # number of groups
+
+    rhashes::Vector{UInt} # row hashes
+    gslots::Vector{Int}   # hashindex -> index of group-representative row
+
+    groups::Vector{Int}   # group index for each row
+    rperm::Vector{Int}    # permutation of row indices that sorts them by groups
+
+    starts::Vector{Int}   # starts of ranges in rperm for each group
+    stops::Vector{Int}    # stops of ranges in rperm for each group
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(v[i], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
```

Comment:
I hadn't seen your last comment. I think the way null values are hashed is OK since either all values are null and their order does not count, or at least one is non-null, and its hashing function will mix the hash of preceding values, making it order-dependent.

Also, the solutions described in my last comment are correct AFAICT (i.e. they shouldn't generate more collisions than the standard `hash`).